src/server/shell/basic_magnification_manager.cpp

```diff
@@ -0,0 +1,385 @@
+/*
+* Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "basic_magnification_manager.h"
+
+#include "mir/synchronised.h"
+#include "mir/compositor/scene_element.h"
+#include "mir/compositor/screen_shooter.h"
+#include "mir/events/event.h"
+#include "mir/events/input_event.h"
+#include "mir/events/pointer_event.h"
+#include "mir/frontend/surface_stack.h"
+#include "mir/graphics/cursor.h"
+#include "mir/graphics/graphic_buffer_allocator.h"
+#include "mir/graphics/renderable.h"
+#include "mir/input/composite_event_filter.h"
+#include "mir/input/scene.h"
+#include "mir/input/seat.h"
+#include "mir/scene/scene_change_notification.h"
+#include "mir/renderer/sw/pixel_source.h"
+
+#include <future>
+
+namespace mg = mir::graphics;
+namespace mi = mir::input;
+namespace mrs = mir::renderer::software;
+namespace ms = mir::scene;
+namespace msh = mir::shell;
+namespace geom = mir::geometry;
+
+namespace
+{
+struct State
+{
+    geom::PointF cursor_position;
+    geom::Point capture_position;
+    bool enabled_ = false;
+};
+
+class DoubleBufferStream {
```

Comment:
I think it's *possible* that just using `mc::Stream` would be usable here? Although you'd need to track buffer usage and such, so maybe what you've done here is better?

Although, maybe you need that anyway? This doesn't have any way of verifying that the read buffer is unused before writing to it, nor does it have a way of verifying that the write buffer has finished being written to before it's read from?